File: leetcode/array/74.SearchA2DMatrix/74Searcha2DMatrix.py
'''遍历所有数据'''
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not len(matrix):
            return False
        n = len(matrix[0])
        left = 0
        right = len(matrix) * n -1
        while left <= right:
            mid = left + (right - left) // 2
            if target == matrix[mid//n][mid%n]:
                return True
            elif target < matrix[mid//n][mid%n]:
                right = mid-1
            else:
                left = mid+1
        return False


######################################
#######222222222222###################
'''遍历一行一列'''
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix or not matrix[0]:
            return False
        m, n = len(matrix), len(matrix[0])
        left, right = 0, m-1
        target_row = None
        while left <= right:
            mid = left + (right - left) // 2
            if matrix[mid][n-1] == target:
                target_row = mid
                break
            elif matrix[mid][n-1] < target:
                left = mid + 1
            else:
                right = mid-1
                
        if target_row == None: 
            target_row = left
        if target_row == m:
            return False
        
        left, right = 0, n-1
        while left <= right:
            mid = left + (right - left) // 2
            if matrix[target_row][mid] == target:
                return True
            elif matrix[target_row][mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return False

File: leetcode/array/74.SearchA2DMatrix/74Searcha2DMatrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */

//  从左下角 向右向上查找
var searchMatrix = function (matrix, target) {
    // matrix[上下][左右]
    let row = matrix.length - 1; // 行
    let column = 0; // 列
    while (row >= 0 && column <= matrix[0].length - 1) {
        if (matrix[row][column] == null) {
            return false;
        }
        if (target > matrix[row][column]) {
            column += 1;
        } else if (target < matrix[row][column]) {
            row -= 1;
        } else if (target === matrix[row][column]) {
            return true;
        }
    }
    return false;
};
